feature-libs/product-configurator/rulebased/root/cpq/interceptor/cpq-access-storage.service.ts

```diff
@@ -0,0 +1,127 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { AuthService } from '@spartacus/core';
+import { BehaviorSubject, Observable, Subscription, timer } from 'rxjs';
+import {
+  distinctUntilChanged,
+  expand,
+  filter,
+  switchMap,
+  take,
+} from 'rxjs/operators';
+import { CpqAccessData } from './cpq-access-data.models';
+import { CpqAccessLoaderService } from './cpq-access-loader.service';
+import { CpqConfiguratorAuthConfig } from './cpq-configurator-auth.config';
+
+@Injectable({ providedIn: 'root' })
+export class CpqAccessStorageService implements OnDestroy {
+  protected readonly EXPIRED_TOKEN: CpqAccessData = {
+    accessToken: 'INVALID DUMMY',
+    accessTokenExpirationTime: 0,
+  };
+
+  constructor(
+    protected cpqAccessLoaderService: CpqAccessLoaderService,
+    protected authService: AuthService,
+    protected config: CpqConfiguratorAuthConfig
+  ) {}
+
+  ngOnDestroy(): void {
+    this.currentCpqAccessSubscription.unsubscribe();
+  }
+
+  protected cpqAccessData$: Observable<CpqAccessData>;
+  protected currentCpqAccessSubscription: Subscription;
+  protected _cpqAccessData$: BehaviorSubject<CpqAccessData>;
+
+  getCpqAccessData(): Observable<CpqAccessData> {
+    if (!this.cpqAccessData$ || this._cpqAccessData$.hasError) {
+      this.initCpqAccessData();
+    }
+    return this.cpqAccessData$;
+  }
+
+  /**
+   * Renews the current access data. All subscribers of getCachedCpqAccessData()
+   * will receive the new data. Will only have an effect, if there are any subscribers
+   * and the user is logged in.
+   */
+  renewCpqAccessData() {
+    // only force token refresh if initialized.
+    if (this.cpqAccessData$) {
+      this.stopAutoFetchingCpqAccessData();
+      this._cpqAccessData$.next(this.EXPIRED_TOKEN); // invalidate cache
+      this.authService
+        .isUserLoggedIn()
+        .pipe(take(1)) // get current login state
+        .subscribe((loggedIn) => {
+          // only fetch new token if user is logged in.
+          if (loggedIn) {
+            this.startAutoFetchingCpqAccessData();
+          }
+        });
+    }
+  }
+
+  protected initCpqAccessData() {
+    this._cpqAccessData$ = new BehaviorSubject(this.EXPIRED_TOKEN);
+    this.cpqAccessData$ = this._cpqAccessData$.pipe(
+      // Never expose expired tokens - either cache was invalidated with expired token,
+      // or the cached one expired before a new one was fetched.
+      filter((data) => !this.isTokenExpired(data))
+    );
+    this.authService
+      .isUserLoggedIn()
+      .pipe(distinctUntilChanged()) // only react if user login status changes
+      .subscribe((loggedIn) => {
```

Comment:
Becasue `initCpqAccessData()` might be invoked many times (in case of error in behaviorSubject), we should:
- maintain a reference to that subscription
- unsubscribe old one before creating new one 

Otherwise we can introduce memory leaks each time the BehaviorSubject gets error in the browser. And even worse: the subscription callback `(loggedIn) => ....` will be called also for hanging subscriptions.

btw. nice to have: we can also unsubscribe in ngOnDestroy. However that's not crucial. Our logic doesn't run in SSR